### PR TITLE
Add colcon related recipes.

### PR DESCRIPTION
**Pull request by seanyen**
Checklist

- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vend other packages
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

### CODE REVIEW

**Comment by seanyen** [comment]
@conda-forge/staged-recipes This is ready for review and merge.

NOTE: `colcon-notification` is disabled for `osx` and `unix` because they depend on `notify2` which requires more packages to be ported and I am not going to tackle them this time.

**Comment by seanyen** [comment]
@wolfv Do you also want to be a co-maintainer of those packages?

**Comment by wolfv** [comment]
Yes, that would be nice

File: recipes/colcon-cmake/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    - colcon-library-path
+    - colcon-test-result >=0.3.3
+
+test:
+  imports:
+    - {{ name.replace("-", "_") }}
+
+about:
+  home: https://github.com/colcon/{{ name }}
+  license: Apache-2.0
```
Comment (chrisburr) [inline recipes/colcon-cmake/meta.yaml]:
```suggestion
  license: Apache-2.0
  license_family: Apache
```

File: recipes/colcon-cmake/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  version: "{{ version }}"
+
+source:
+  url: https://github.com/colcon/{{ name }}/archive/{{ version }}.tar.gz
+  sha256: 5f87370765b10cccfca2117e32cf194361703a478d88b34037f94ca76c1d3189
+
+build:
+  number: 0
+  script: "{{ PYTHON }} -m pip install . --no-deps --ignore-installed -vv "
+  skip: True  # [py<35]
```
Comment (chrisburr) [inline recipes/colcon-cmake/meta.yaml]: For those that have the same dependencies on all platforms
```suggestion
  noarch: python
```

File: recipes/colcon-cmake/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+requirements:
+  host:
+    - pip
+    - python
```
Comment (chrisburr) [inline recipes/colcon-cmake/meta.yaml]: For those that become `noarch: python`:
```suggestion
    - python >=3.6
```

File: recipes/colcon-cmake/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  run:
+    - python
```
Comment (chrisburr) [inline recipes/colcon-cmake/meta.yaml]: For those that become `noarch: python`:
```suggestion
    - python >=3.6
```

**Comment by seanyen** [comment]
@chrisburr Good suggestions. I didn't think of using them. Thanks!

**Comment by seanyen** [comment]
@conda-forge/staged-recipes Help! On the Windows `conda` build, it fails at a rename step when packing the package. It seems trying to create an invalid folder name like `work_moved_colcon-core-0.4.0-py>=35_0_win-64_main_build_loop` and the error manifests as below:

Is it something known? Any workaround?

```
Renaming work directory,  C:\Miniconda\conda-bld\colcon-core_1570769496978\work  to  C:\Miniconda\conda-bld\colcon-core_1570769496978\work_moved_colcon-core-0.4.0-py>=35_0_win-64_main_build_loop
Traceback (most recent call last):
  File "C:\Miniconda\lib\shutil.py", line 544, in move
    os.rename(src, real_dst)
OSError: [WinError 123] The filename, directory name, or volume label syntax is incorrect: 'C:\\Miniconda\\conda-bld\\colcon-core_1570769496978\\work' -> 'C:\\Miniconda\\conda-bld\\colcon-core_1570769496978\\work_moved_colcon-core-0.4.0-py>=35_0_win-64_main_build_loop'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File ".ci_support\build_all.py", line 134, in <module>
    build_all(args.recipes_dir, args.arch)
  File ".ci_support\build_all.py", line 68, in build_all
    build_folders(recipes_dir, new_comp_folders, arch, channel_urls)
  File ".ci_support\build_all.py", line 124, in build_folders
    conda_build.api.build([recipe], config=get_config(arch, channel_urls))
  File "C:\Miniconda\lib\site-packages\conda_build\api.py", line 209, in build
    notest=notest, need_source_download=need_source_download, variants=variants)
  File "C:\Miniconda\lib\site-packages\conda_build\build.py", line 2408, in build_tree
    shutil.move(metadata.config.work_dir, dest)
  File "C:\Miniconda\lib\shutil.py", line 555, in move
    symlinks=True)
  File "C:\Miniconda\lib\shutil.py", line 315, in copytree
    os.makedirs(dst)
  File "C:\Miniconda\lib\os.py", line 220, in makedirs
    mkdir(name, mode)
OSError: [WinError 123] The filename, directory name, or volume label syntax is incorrect: 'C:\\Miniconda\\conda-bld\\colcon-core_1570769496978\\work_moved_colcon-core-0.4.0-py>=35_0_win-64_main_build_loop'

```

File: recipes/colcon-common-extensions/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  sha256: a8d78f5f01f99d1d6f0bf490427281773dab7ff046ba4e746c80816c9419dda6
+
+build:
+  number: 0
+  script: "{{ PYTHON }} -m pip install . --no-deps --ignore-installed -vv "
+
+requirements:
+  host:
+    - pip
+    - python >=3.5
```
Comment (chrisburr) [inline recipes/colcon-common-extensions/meta.yaml]: This syntax can only be used if the package is noarch python. If it’s not you have to use skip true py<35 way.

Comment (chrisburr) [reply recipes/colcon-common-extensions/meta.yaml]: To add, the issue with doing it this way is that you have one build for each supported python version and python =2.7 conflicts with python >=3.5.

File: recipes/colcon-core/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  script: "{{ PYTHON }} -m pip install . --no-deps --ignore-installed -vv "
+  entry_points:
+    - colcon = colcon_core.command:main
+
+requirements:
+  host:
+    - pip
+    - python {{ python }}
+  run:
+    - python >=3.5
```
Comment (chrisburr) [inline recipes/colcon-core/meta.yaml]: `conda_build_config.yaml` shouldn't be used at all. The issue is that you only changed the `host` requirement. The `run` requirement should be changed to match.

The pattern is either:
```yaml
build:
  noarch: python

requirements:
  host:
    - python >=3.5
  run:
    - python >=3.5
    - dependency_that_is_for_all_platforms_and_versions
```
or:
```yaml
build:
  skip: true  # [py<35]

requirements:
  host:
    - python >=3.5
  run:
    - python >=3.5
    - some_windows_only_dependency  # [win]
```

Comment (seanyen) [reply recipes/colcon-core/meta.yaml]: As long as I am specifying the `python >=3.5` under `host`. The renaming error will manifest.

It seems only the below combination works for me locally. :(
(Trigger another CI build to confirm other platforms...)

```yaml
build:
  skip: true  # [py<35]

requirements:
  host:
    - python
  run:
    - python >=3.5
    - some_windows_only_dependency  # [win]
```

Comment (chrisburr) [reply recipes/colcon-core/meta.yaml]: Sorry there was a mistake in the second code block. I've corrected it now.

Comment (seanyen) [reply recipes/colcon-core/meta.yaml]: No problem! And finally CI shows green again. :)

**Comment by wolfv** [comment]
wow this is pretty awesome!

**Comment by seanyen** [comment]
cc @conda-forge/staged-recipes This is ready for review and merge. Thanks!

File: .circleci/build_steps.sh [hunk 7]
```diff
@@ -43,6 +43,7 @@ export CONDA_NPY='19'
 conda index /home/conda/staged-recipes/build_artifacts
 
 conda install --yes --quiet "conda!=4.6.1,<4.7.11a0" conda-forge-ci-setup=2.* conda-forge-pinning networkx "conda-build>=3.16"
+export FEEDSTOCK_ROOT="${FEEDSTOCK_ROOT:-/home/conda/staged-recipes}"
```
Comment (isuruf) [inline .circleci/build_steps.sh]: Thanks @chrisburr

**Pull request merged**